integration_tests/README.md

```diff
@@ -303,3 +303,94 @@ The marks you care about are all in marks.py
 
 For the most part you can ignore this file. It provides the underlying Spark session to operations that need it, but most tests should interact with
 it through `asserts.py`.
+
+## Guidelines for Testing
+
+When support for a new operator is added to `rapids-plugin-4-spark`, it is recommended that
+the following conditions be covered in its corresponding integration tests:
+
+### 1. Cover all supported data types
+Ensure that tests cover all data types supported by the added operation. An exhaustive list of data types supported in 
+Spark is available [here](https://spark.apache.org/docs/latest/sql-ref-datatypes.html). These include:
+   * Numeric Types 
+     * `ByteType` 
+     * `ShortType` 
+     * `IntegerType`
+     * `LongType`
+     * `FloatType`
+     * `DoubleType`
+     * `DecimalType`
+   * Strings 
+     * `StringType` 
+     * `VarcharType`
+   * Binary (`BinaryType`)  
+   * Booleans (`BooleanType`)
+   * Chrono Types 
+     * `TimestampType` 
+     * `DateType`
+     * `Interval`
+   * Complex Types 
+     * `ArrayType`
+     * `StructType`
+     * `MapType`
+
+`data_gen.py` provides `DataGen` classes that help generate test data in integration tests.
+
+### 2. Nested Data types
+Complex data types (`ArrayType`, `StructType`, `MapType`) warrant extensive testing for various combinations of nesting.
+E.g.
+   * `Array<primitive_type>`
+   * `Array<Array<primitive_type>>`
+   * `Array<Struct<primitive_type>>`
+   * `Struct<Array<primitive_type>>`
+   * `Array<Struct<Array<primitive_type>>>`
+   * `Struct<Array<Struct<primitive_type>>>`
+
+The `ArrayGen` and `StructGen` classes in `data_gen.py` can be configured to support arbitrary nesting.
+
+### 3. Scalar values
+Operators and expressions that support scalar operands need to be tested with scalar inputs, of all 
+supported types from 1 and 2, above. 
+For instance, `SUM()` supports numeric columns (e.g. `SUM(a + b)`), or scalars (e.g. `SUM(20)`).
+Similarly, `COUNT()` supports the following:
+   * Columns: E.g. `COUNT(a)` to count non-null rows for column `a`
+   * Scalars: E.g. `COUNT(1)` to count all rows (including nulls)
+   * `*`: E.g. `COUNT(*)`, functionally equivalent to `COUNT(1)`
+It is advised that tests be added for all applicable scalar types, for an operator.
+
+### 4. Null rows
+Ensure that the test data accommodates null values for input columns.
+Null values in input columns are a frequent source of bugs in `rapids-plugin-4-spark`, because of mismatches in null-handling
+and semantics, between RAPIDS `libcudf` (on which `rapids-plugin-4-spark` relies heavily), and Apache Spark. 
+
+Apart from null rows in columns of primitive types, the following conditions must be covered for nested types:
+   * Null rows at the "top" level for `Array`/`Struct` columns.   E.g. `[ [1,2], [3], ∅, [4,5,6] ]`.
+   * Non-null rows containing null elements in the child column. E.g. `[ [1,2], [3,∅], ∅, [4,∅,6] ]`.
+
```

Comment:
1) It would be good to point out that the data gen inserts nulls by default, but this can be over-loaded
2) It would be good to point out that null literal values are also important to test.
3) It might be good to point out that for all aggregations, including group by, reductions, and windowing it is good to test where there all of the data being processed are nulls.

Reply:
Good one. Added.